In [3]:
# 필요 모듈

import pandas as pd
import numpy as np
import warnings;warnings.filterwarnings('ignore')

## 데이터 확인 및 정제

In [5]:
movies = pd.read_csv('/content/drive/MyDrive/Colab Notebooks/머신러닝 완벽 가이드/00_데이터셋/tmdb_5000_movies.csv')
print(movies.shape)
movies.head()


(4803, 20)


budget                                             genres  \
0  237000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
1  300000000  [{"id": 12, "name": "Adventure"}, {"id": 14, "...   
2  245000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   
3  250000000  [{"id": 28, "name": "Action"}, {"id": 80, "nam...   
4  260000000  [{"id": 28, "name": "Action"}, {"id": 12, "nam...   

                                       homepage      id  \
0                   http://www.avatarmovie.com/   19995   
1  http://disney.go.com/disneypictures/pirates/     285   
2   http://www.sonypictures.com/movies/spectre/  206647   
3            http://www.thedarkknightrises.com/   49026   
4          http://movies.disney.com/john-carter   49529   

                                            keywords original_language  \
0  [{"id": 1463, "name": "culture clash"}, {"id":...                en   
1  [{"id": 270, "name": "ocean"}, {"id": 726, "na...                en   
2  [{"id": 470, "name": "spy"}, {"id": 818, "name...                en   
3  [{"id": 849, "name": "dc comics"}, {"id": 853,...                en   
4  [{"id": 818, "name": "based on novel"}, {"id":...                en   

                             original_title  \
0                                    Avatar   
1  Pirates of the Caribbean: At World's End   
2                                   Spectre   
3                     The Dark Knight Rises   
4                               John Carter   

                                            overview  popularity  \
0  In the 22nd century, a paraplegic Marine is di...  150.437577   
1  Captain Barbossa, long believed to be dead, ha...  139.082615   
2  A cryptic message from Bond’s past sends him o...  107.376788   
3  Following the death of District Attorney Harve...  112.312950   
4  John Carter is a war-weary, former military ca...   43.926995   

                                production_companies  \
0  [{"name": "Ingenious Film Partners", "id": 289...   
1  [{"name": "Walt Disney Pictures", "id": 2}, {"...   
2  [{"name": "Columbia Pictures", "id": 5}, {"nam...   
3  [{"name": "Legendary Pictures", "id": 923}, {"...   
4        [{"name": "Walt Disney Pictures", "id": 2}]   

                                production_countries release_date     revenue  \
0  [{"iso_3166_1": "US", "name": "United States o...   2009-12-10  2787965087   
1  [{"iso_3166_1": "US", "name": "United States o...   2007-05-19   961000000   
2  [{"iso_3166_1": "GB", "name": "United Kingdom"...   2015-10-26   880674609   
3  [{"iso_3166_1": "US", "name": "United States o...   2012-07-16  1084939099   
4  [{"iso_3166_1": "US", "name": "United States o...   2012-03-07   284139100   

   runtime                                   spoken_languages    status  \
0    162.0  [{"iso_639_1": "en", "name": "English"}, {"iso...  Released   
1    169.0           [{"iso_639_1": "en", "name": "English"}]  Released   
2    148.0  [{"iso_639_1": "fr", "name": "Fran\u00e7ais"},...  Released   
3    165.0           [{"iso_639_1": "en", "name": "English"}]  Released   
4    132.0           [{"iso_639_1": "en", "name": "English"}]  Released   

                                          tagline  \
0                     Enter the World of Pandora.   
1  At the end of the world, the adventure begins.   
2                           A Plan No One Escapes   
3                                 The Legend Ends   
4            Lost in our world, found in another.   

                                      title  vote_average  vote_count  
0                                    Avatar           7.2       11800  
1  Pirates of the Caribbean: At World's End           6.9        4500  
2                                   Spectre           6.3        4466  
3                     The Dark Knight Rises           7.6        9106  
4                               John Carter           6.1        2124

        tmdbsms 4803의 레코드와 20개의 피처로 구성
        (영화 제목, 개요, 인기도, 평점, 투표수, 예산, 키워드 등) 영화에 대한 메타 정보

        => 콘텐츠 기반 필터링 추천 분석에 사용할 주요 컬럼
        (id, 영화제목, 장르, 평균 평점, 평점 투표수, 영화 인기도, 영화 키워드, 개요)

In [6]:
movies_df = movies[['id', 'title', 'genres', 'vote_average', 'vote_count', 'popularity', 'keywords', 'overview']]
movies_df.head()

id                                     title  \
0   19995                                    Avatar   
1     285  Pirates of the Caribbean: At World's End   
2  206647                                   Spectre   
3   49026                     The Dark Knight Rises   
4   49529                               John Carter   

                                              genres  vote_average  \
0  [{"id": 28, "name": "Action"}, {"id": 12, "nam...           7.2   
1  [{"id": 12, "name": "Adventure"}, {"id": 14, "...           6.9   
2  [{"id": 28, "name": "Action"}, {"id": 12, "nam...           6.3   
3  [{"id": 28, "name": "Action"}, {"id": 80, "nam...           7.6   
4  [{"id": 28, "name": "Action"}, {"id": 12, "nam...           6.1   

   vote_count  popularity                                           keywords  \
0       11800  150.437577  [{"id": 1463, "name": "culture clash"}, {"id":...   
1        4500  139.082615  [{"id": 270, "name": "ocean"}, {"id": 726, "na...   
2        4466  107.376788  [{"id": 470, "name": "spy"}, {"id": 818, "name...   
3        9106  112.312950  [{"id": 849, "name": "dc comics"}, {"id": 853,...   
4        2124   43.926995  [{"id": 818, "name": "based on novel"}, {"id":...   

                                            overview  
0  In the 22nd century, a paraplegic Marine is di...  
1  Captain Barbossa, long believed to be dead, ha...  
2  A cryptic message from Bond’s past sends him o...  
3  Following the death of District Attorney Harve...  
4  John Carter is a war-weary, former military ca...

In [7]:
## 데이터 확인하기 

# 데이터의 구성

movies_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4803 entries, 0 to 4802
Data columns (total 8 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   id            4803 non-null   int64  
 1   title         4803 non-null   object 
 2   genres        4803 non-null   object 
 3   vote_average  4803 non-null   float64
 4   vote_count    4803 non-null   int64  
 5   popularity    4803 non-null   float64
 6   keywords      4803 non-null   object 
 7   overview      4800 non-null   object 
dtypes: float64(2), int64(2), object(4)
memory usage: 300.3+ KB


In [ ]:
# Genres , keywords 의 경우 딕셔너리의 형태를 지닌 문자열 => 개별 장르를 파이썬 리스트 객체로 추철

from ast import literal_eval


## 문자열을 객체로 변환!
movies_df['genres'] = movies_df['genres'].apply(literal_eval)
movies_df['keywords'] = movies_df['keywords'].apply(literal_eval)


In [14]:
movies_df[['genres', 'keywords']][:1]

genres  \
0  [{'id': 28, 'name': 'Action'}, {'id': 12, 'nam...   

                                            keywords  
0  [{'id': 1463, 'name': 'culture clash'}, {'id':...

In [15]:
## 딕셔너리의 name 의 values 들을 list 의 형태로 저장 

movies_df['genres'] = movies_df['genres'].apply(lambda x : [y['name'] for y in x])
movies_df['keywords'] = movies_df['keywords'].apply(lambda x : [y['name'] for y in x])

In [16]:
movies_df[['genres', 'keywords']][:1]

genres  \
0  [Action, Adventure, Fantasy, Science Fiction]   

                                            keywords  
0  [culture clash, future, space war, space colon...

## 장르 콘텐츠의 유사도 측정

        genres 를 문자열로 변경한 뒤 이를 CountVectorizer로 피처 벡터화한 행렬 데이터 값을 코사인 유사도로 비교

- 문자열로 변환된 genres 칼럼을 Count 기반으로 피처 벡터화 변환
- genres 문자열을 피처 벡터화 행렬로 변환한 데이터 세트를 _**코사인 유사도**_를 통해 비교  
레코드별로 타 레코드와 장르에서 코사인 유사도 값을 가지는 객체를 생성
- 장르 유사도가 높은 영화 중 평점이 높은 순으로 추천


* CountVectorizer \
: 단어들의 카운트(출현 빈도(frequency))로 여러 문서들을 벡터화\
카운트 행렬, 단어 문서 행렬 (Term-Document Matrix, TDM))\
모두 소문자로 변환시키기 때문에 me 와 Me 는 모두 같은 특성이 된다.

In [19]:
# CounterVectorizer : 피처 벡터 행렬


from sklearn.feature_extraction.text import CountVectorizer

# genres 의 리스트를 공배으로 나뉘어진 문자열로 변환

movies_df['genres_literal'] = movies_df['genres'].apply(lambda x : (' ').join(x))

count_vect = CountVectorizer(min_df = 0 , ngram_range=(1, 2))
genre_mat = count_vect.fit_transform(movies_df['genres_literal'])
print(genre_mat.shape)


(4803, 276)


<4803x276 sparse matrix of type '<class 'numpy.int64'>'
	with 20631 stored elements in Compressed Sparse Row format>

In [21]:
# 코사인 유사도를 통해 비교

from sklearn.metrics.pairwise import cosine_similarity

genre_sim = cosine_similarity(genre_mat, genre_mat)
print(genre_sim.shape)
print(genre_sim[0].shape)
print(genre_sim[2].shape)
print(genre_sim[:2])
## 행별 장르 유사도 값을 가지고 있는것 
### 1, 2 => 1행과 2행의 장르 유사도

(4803, 4803)
(4803,)
(4803,)
[[1.         0.59628479 0.4472136  ... 0.         0.         0.        ]
 [0.59628479 1.         0.4        ... 0.         0.         0.        ]]


In [29]:
# 유사도가 높은 순으로 다른 레코드를 추춘

genre_sim_soreted_ind = genre_sim.argsort()[:, ::-1]
print(genre_sim_soreted_ind)

# 각 행별 유사도가 높은 순서대로 추출

## 1행을 보자면 => 가장 높은 유사도는 당연 자기 자신을 제외하고, 3494을 의미한다. 가장 낮은 유사도는 2401

[[   0 3494  813 ... 3038 3037 2401]
 [ 262    1  129 ... 3069 3067 2401]
 [   2 1740 1542 ... 3000 2999 2401]
 ...
 [4800 3809 1895 ... 2229 2230    0]
 [4802 1594 1596 ... 3204 3205    0]
 [4802 4710 4521 ... 3140 3141    0]]


### 장르 콘텐츠 필터링을 이용한 영화 추천

In [30]:
## 장르 유사도에 따라 영화 추천


def find_sim_movie(df, sorted_ind, title_name, top_n=10):
    # df 중 title 이 같은 것만 추출
    title_movie = df[df['title'] == title_name]

    title_index = title_movie.index.values
    similar_indexes = sorted_ind[title_index, :(top_n)]

    print(similar_indexes)
    similar_indexes = similar_indexes.reshape(-1)

    return df.iloc[similar_indexes]



similar_movies = find_sim_movie(movies_df, genre_sim_soreted_ind, 'The Godfather', 10)
similar_movies


[[2731 1243 3636 1946 2640 4065 1847 4217  883 3866]]


id                                           title          genres  \
2731     240                          The Godfather: Part II  [Drama, Crime]   
1243     203                                    Mean Streets  [Drama, Crime]   
3636   36351                                   Light Sleeper  [Drama, Crime]   
1946   11699  The Bad Lieutenant: Port of Call - New Orleans  [Drama, Crime]   
2640     400         Things to Do in Denver When You're Dead  [Drama, Crime]   
4065  364083                                      Mi America  [Drama, Crime]   
1847     769                                      GoodFellas  [Drama, Crime]   
4217    9344                                            Kids  [Drama, Crime]   
883      640                             Catch Me If You Can  [Drama, Crime]   
3866     598                                     City of God  [Drama, Crime]   

      vote_average  vote_count  popularity  \
2731           8.3        3338  105.792936   
1243           7.2         345   17.002096   
3636           5.7          15    6.063868   
1946           6.0         326   17.339852   
2640           6.7          85    6.932221   
4065           0.0           0    0.039007   
1847           8.2        3128   63.654244   
4217           6.8         279   13.291991   
883            7.7        3795   73.944049   
3866           8.1        1814   44.356711   

                                               keywords  \
2731  [italo-american, cuba, vororte, melancholy, pr...   
1243  [epilepsy, protection money, secret love, mone...   
3636  [suicide, drug dealer, redemption, addict, exi...   
1946  [police brutality, organized crime, policeman,...   
2640  [father son relationship, bounty hunter, boat,...   
4065                       [new york state, hate crime]   
1847  [prison, based on novel, florida, 1970s, mass ...   
4217                              [puberty, first time]   
883   [con man, biography, fbi agent, overhead camer...   
3866  [male nudity, street gang, brazilian, photogra...   

                                               overview genres_literal  
2731  In the continuing saga of the Corleone crime f...    Drama Crime  
1243  A small-time hood must choose from among love,...    Drama Crime  
3636  A drug dealer with upscale clientele is having...    Drama Crime  
1946  Terrence McDonagh, a New Orleans Police sergea...    Drama Crime  
2640  A mafia film in Tarantino style with a star-st...    Drama Crime  
4065  A hate-crime has been committed in a the small...    Drama Crime  
1847  The true story of Henry Hill, a half-Irish, ha...    Drama Crime  
4217  A controversial portrayal of teens in New York...    Drama Crime  
883   A true story about Frank Abagnale Jr. who, bef...    Drama Crime  
3866  Cidade de Deus is a shantytown that started du...    Drama Crime

In [32]:
##평점 높은 순으로 단, 평가자가 1명인 경우를 고려해야 하므로 가중평점을 이요하자 


C = movies_df['vote_average'].mean()
m = movies_df['vote_count'].quantile(0.6) # 투표 횟수가 많을 수록 가중치 부여 => 상위 60%에게 가중치 부여

print("C :", round(C, 3), 'm : ', round(m, 3))

C : 6.092 m :  370.2


In [35]:
# 가중치를 부여할 상위 percent

percentile = 0.6

C = movies_df['vote_average'].mean()
m = movies_df['vote_count'].quantile(percentile)

def weighted_vote_average(record):
    v = record['vote_count']
    R = record['vote_average']

    return ( (v/(v+m)) * R ) + ( (m/(m+v)) * C )

movies_df['weight_vote'] = movies_df.apply(weighted_vote_average, axis = 1)




In [38]:
movies_df[['title', 'vote_average', 'weight_vote', 'vote_count']].sort_values('weight_vote', ascending=False)[:10]

title  vote_average  weight_vote  vote_count
1881  The Shawshank Redemption           8.5     8.396052        8205
3337             The Godfather           8.4     8.263591        5893
662                 Fight Club           8.3     8.216455        9413
3232              Pulp Fiction           8.3     8.207102        8428
65             The Dark Knight           8.2     8.136930       12002
1818          Schindler's List           8.3     8.126069        4329
3865                  Whiplash           8.3     8.123248        4254
809               Forrest Gump           8.2     8.105954        7927
2294             Spirited Away           8.3     8.105867        3840
2731    The Godfather: Part II           8.3     8.079586        3338

In [39]:
## 종합, 장르 유사도 높고 평점 높은 순으로 추천



def find_sim_movie(df, sorted_ind, title_name, top_n=10):
    # df 중 title 이 같은 것만 추출
    title_movie = df[df['title'] == title_name]

    title_index = title_movie.index.values
    similar_indexes = sorted_ind[title_index, :(top_n)*2]
    similar_indexes = similar_indexes[similar_indexes != title_index] # 자기 자신 제외


    print(similar_indexes)
    similar_indexes = similar_indexes.reshape(-1)

    return df.iloc[similar_indexes].sort_values('weight_vote', ascending=False)[:top_n]




similar_movies = find_sim_movie(movies_df, genre_sim_soreted_ind, 'The Godfather', 10)
similar_movies

[2731 1243 3636 1946 2640 4065 1847 4217  883 3866 3112 4041  588 3378
  281 1663 1464 1149 2839]


id                        title          genres  vote_average  \
2731     240       The Godfather: Part II  [Drama, Crime]           8.3   
1847     769                   GoodFellas  [Drama, Crime]           8.2   
3866     598                  City of God  [Drama, Crime]           8.1   
1663     311  Once Upon a Time in America  [Drama, Crime]           8.2   
883      640          Catch Me If You Can  [Drama, Crime]           7.7   
281     4982            American Gangster  [Drama, Crime]           7.4   
4041   11798              This Is England  [Drama, Crime]           7.4   
1149  168672              American Hustle  [Drama, Crime]           6.8   
1243     203                 Mean Streets  [Drama, Crime]           7.2   
2839   10220                     Rounders  [Drama, Crime]           6.9   

      vote_count  popularity  \
2731        3338  105.792936   
1847        3128   63.654244   
3866        1814   44.356711   
1663        1069   49.336397   
883         3795   73.944049   
281         1502   42.361215   
4041         363    8.395624   
1149        2807   49.664128   
1243         345   17.002096   
2839         439   18.422008   

                                               keywords  \
2731  [italo-american, cuba, vororte, melancholy, pr...   
1847  [prison, based on novel, florida, 1970s, mass ...   
3866  [male nudity, street gang, brazilian, photogra...   
1663  [life and death, corruption, street gang, rape...   
883   [con man, biography, fbi agent, overhead camer...   
281   [underdog, black people, drug traffic, drug sm...   
4041  [holiday, skinhead, england, vandalism, indepe...   
1149             [con artist, scam, mobster, fbi agent]   
1243  [epilepsy, protection money, secret love, mone...   
2839  [gambling, law, compulsive gambling, roulette,...   

                                               overview genres_literal  \
2731  In the continuing saga of the Corleone crime f...    Drama Crime   
1847  The true story of Henry Hill, a half-Irish, ha...    Drama Crime   
3866  Cidade de Deus is a shantytown that started du...    Drama Crime   
1663  A former Prohibition-era Jewish gangster retur...    Drama Crime   
883   A true story about Frank Abagnale Jr. who, bef...    Drama Crime   
281   Following the death of his employer and mentor...    Drama Crime   
4041  A story about a troubled boy growing up in Eng...    Drama Crime   
1149  A con man, Irving Rosenfeld, along with his se...    Drama Crime   
1243  A small-time hood must choose from among love,...    Drama Crime   
2839  A young man is a reformed gambler who must ret...    Drama Crime   

      weight_vote  
2731     8.079586  
1847     7.976937  
3866     7.759693  
1663     7.657811  
883      7.557097  
281      7.141396  
4041     6.739664  
1149     6.717525  
1243     6.626569  
2839     6.530427